File: internal/httpapi/conformance_test.go
package httpapi_test

// Conformance runner: replays spec/conformance/cases/*.json against the HTTP
// handler built by httpapi.New, the same constructor cmd/tunna uses.
//
// Requests are signed with package sig using the keys in fixtures.json.
// Every case gets a fresh server with every fixture provisioned (keys,
// buckets, objects with bytes in a real disk store), so cases cannot affect
// each other and a case may mutate a fixture freely. Metadata lives in the
// memory stores here: this suite tests the wire; persistence has its own
// tests in internal/sqlite and cmd/tunna.

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"net/url"
	"os"
	"path/filepath"
	"reflect"
	"strconv"
	"strings"
	"testing"
	"time"

	"github.com/tunnaio/tunna/internal/disk"
	"github.com/tunnaio/tunna/internal/fixtures"
	"github.com/tunnaio/tunna/internal/httpapi"
	"github.com/tunnaio/tunna/internal/memory"
	"github.com/tunnaio/tunna/sig"
)

const specDir = "../../spec"

// --- spec file shapes (subset of the schemas, enough to run) ---

type caseFile struct {
	Spec  string            `json:"spec"`
	Cases []conformanceCase `json:"cases"`
}

type conformanceCase struct {
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Fixtures    *[]string `json:"fixtures"` // nil means "all"
	Steps       []step    `json:"steps"`
}

type step struct {
	Name    string            `json:"name"`
	Request request           `json:"request"`
	Expect  expect            `json:"expect"`
	Capture map[string]string `json:"capture"`
}

type request struct {
	Method         string            `json:"method"`
	Path           []string          `json:"path"`
	Query          [][2]string       `json:"query"`
	Headers        map[string]string `json:"headers"`
	Auth           json.RawMessage   `json:"auth"`
	Body           *body             `json:"body"`
	ExpectContinue bool              `json:"expect_continue"`
}

type body struct {
	JSON  json.RawMessage `json:"json"`
	Text  *string         `json:"text"`
	Hex   *string         `json:"hex"`
	Bytes *genBytes       `json:"bytes"`
}

type genBytes struct {
	Seed   int64 `json:"seed"`
	Length int64 `json:"length"`
}

type expect struct {
	Status      *int                       `json:"status"`
	Error       *string                    `json:"error"`
	Headers     map[string]json.RawMessage `json:"headers"`
	JSON        json.RawMessage            `json:"json"`
	JSONPresent []string                   `json:"json_present"`
	JSONAbsent  []string                   `json:"json_absent"`
	BodySHA256  *string                    `json:"body_sha256"`
	BodyLength  *int64                     `json:"body_length"`
	BodyEmpty   *bool                      `json:"body_empty"`
	BodyBytes   *genBytes                  `json:"body_bytes"`
}

type errorTable struct {
	Errors []struct {
		Code   string `json:"code"`
		Status int    `json:"status"`
	} `json:"errors"`
}

// auth is the decoded form of a step's "auth" field.
type auth struct {
	None bool
	Key  *struct {
		Key               string   `json:"key"`
		SignedHeaders     []string `json:"signed_headers"`
		TimeOffsetSeconds int64    `json:"time_offset_seconds"`
		CorruptSignature  bool     `json:"corrupt_signature"`
	}
	Presign *struct {
		Key              string   `json:"key"`
		ExpiresInSeconds int64    `json:"expires_in_seconds"`
		SignedHeaders    []string `json:"signed_headers"`
	}
}

func parseAuth(raw json.RawMessage) (auth, error) {
	var a auth
	if len(raw) == 0 || string(raw) == `"none"` {
		a.None = true
		return a, nil
	}
	var probe struct {
		Key     *string         `json:"key"`
		Presign json.RawMessage `json:"presign"`
	}
	if err := json.Unmarshal(raw, &probe); err != nil {
		return a, err
	}
	if probe.Presign != nil {
		return a, json.Unmarshal(probe.Presign, &a.Presign)
	}
	if probe.Key != nil {
		return a, json.Unmarshal(raw, &a.Key)
	}
	return a, fmt.Errorf("unrecognised auth %s", raw)
}

// --- the runner ---

func TestConformance(t *testing.T) {
	version := strings.TrimSpace(mustRead(t, filepath.Join(specDir, "VERSION")))
	statusOf := loadErrorTable(t)
	fx := loadFixtures(t)

	files, err := filepath.Glob(filepath.Join(specDir, "conformance", "cases", "*.json"))
	if err != nil || len(files) == 0 {
		t.Fatalf("no case files under %s: %v", specDir, err)
	}

	for _, file := range files {
		var cf caseFile
		if err := json.Unmarshal([]byte(mustRead(t, file)), &cf); err != nil {
			t.Fatalf("%s: %v", file, err)
		}
		if cf.Spec != version {
			t.Errorf("%s declares spec %q, spec/VERSION is %q", filepath.Base(file), cf.Spec, version)
		}
		for _, c := range cf.Cases {
			c := c
			t.Run(c.Name, func(t *testing.T) {
				if reason := unsupported(c, fx); reason != "" {
					t.Skip(reason)
				}
				// A fresh server per case: fixtures provisioned as fixtures.json
				// describes them, so no case can poison another through shared
				// state, and a case that mutates a fixture is still a valid case.
				srv := newServer(t, fx)
				defer srv.Close()
				runCase(t, srv, c, statusOf, fx)
			})
		}
	}
}

// newServer builds the server under test with every fixture provisioned:
// keys (disabled ones included), buckets, and objects with their bytes in a
// real disk store. Memory stores for metadata, since the conformance suite
// tests the wire, not persistence; the SQLite adapter has its own contract
// tests and cmd/tunna has the restart tests.
func newServer(t *testing.T, fx fixtures.File) *httptest.Server {
	t.Helper()
	blobs, err := disk.New(t.TempDir())
	if err != nil {
		t.Fatalf("disk.New: %v", err)
	}
	now := time.Now()
	objs, err := fx.ObjectRecords(context.Background(), blobs, now)
	if err != nil {
		t.Fatalf("provisioning objects: %v", err)
	}
	objects := memory.NewObjectStore(objs)
	origins, err := fx.Origins()
	if err != nil {
		t.Fatalf("fixtures server.cors_origins: %v", err)
	}
	return httptest.NewServer(httpapi.New(httpapi.Options{
		ServerVersion: "test",
		CORSOrigins:   origins,
		Keys:          memory.NewKeyStore(fx.APIKeys(now)),
		Buckets:       memory.NewBucketStore(fx.BucketRecords(now)),
		Objects:       objects,
		Uploads:       memory.NewUploadStore(nil, objects),
		Blobs:         blobs,
	}))
}

// unsupported returns a reason to skip a case the harness cannot serve.
// Every fixture kind is provisioned into the server before any case runs,
// so the only reason left is a fixture name the file does not define.
func unsupported(c conformanceCase, fx fixtures.File) string {
	if c.Fixtures == nil {
		return ""
	}
	for _, name := range *c.Fixtures {
		if !fx.Has(name) {
			return "unknown fixture " + name
		}
	}
	return ""
}

func runCase(t *testing.T, srv *httptest.Server, c conformanceCase, statusOf map[string]int, fx fixtures.File) {
	captured := map[string]string{}
	numbers := map[string]string{} // captures that were JSON numbers
	for i, s := range c.Steps {
		name := s.Name
		if name == "" {
			name = fmt.Sprintf("step-%d", i+1)
		}
		sub := func(in string) string { return substitute(in, captured) }
		// subJSON is sub for JSON text. A captured number keeps its type
		// where the placeholder is the whole string: "{{name}}" becomes the
		// bare number, in a request body as in an expectation.
		subJSON := func(in string) string {
			for k, v := range numbers {
				in = strings.ReplaceAll(in, `"{{`+k+`}}"`, v)
			}
			return sub(in)
		}

		// Decoded inputs, after substitution.
		segments := make([]string, len(s.Request.Path))
		for i, seg := range s.Request.Path {
			segments[i] = sub(seg)
		}
		query := url.Values{}
		for _, kv := range s.Request.Query {
			query.Add(sub(kv[0]), sub(kv[1]))
		}
		headers := map[string]string{}
		for k, v := range s.Request.Headers {
			headers[k] = sub(v)
		}

		a, err := parseAuth(s.Request.Auth)
		if err != nil {
			t.Fatalf("%s: %v", name, err)
		}

		// URL: sig's encoder is the one under test for the path and query,
		// but the runner must not depend on it being right, so the runner
		// encodes with its own copy.
		var path strings.Builder
		if len(segments) == 0 {
			path.WriteString("/")
		}
		for _, seg := range segments {
			path.WriteString("/")
			path.WriteString(encode(seg))
		}
		requestURL := srv.URL + path.String()

		sigReq := sig.Request{Method: s.Request.Method, Path: segments, Query: query, Headers: headers}
		switch {
		case a.Presign != nil:
			key := fixtureKey(t, fx, a.Presign.Key)
			sigReq.SignedHeaders = a.Presign.SignedHeaders
			expires := time.Now().Unix() + a.Presign.ExpiresInSeconds
			requestURL += "?" + sig.PresignQuery(sigReq, key, expires)
		default:
			var parts []string
			for name, values := range query {
				for _, v := range values {
					parts = append(parts, encode(name)+"="+encode(v))
				}
			}
			if len(parts) > 0 {
				requestURL += "?" + strings.Join(parts, "&")
			}
		}

		// Body
		var reqBody io.Reader
		if b := s.Request.Body; b != nil {
			switch {
			case b.JSON != nil:
				reqBody = strings.NewReader(subJSON(string(b.JSON)))
			case b.Text != nil:
				reqBody = strings.NewReader(sub(*b.Text))
			case b.Hex != nil:
				raw, err := hex.DecodeString(*b.Hex)
				if err != nil {
					t.Fatalf("%s: bad hex body: %v", name, err)
				}
				reqBody = bytes.NewReader(raw)
			case b.Bytes != nil:
				reqBody = bytes.NewReader(fixtures.Generate(b.Bytes.Seed, b.Bytes.Length))
			}
		}

		req, err := http.NewRequest(s.Request.Method, requestURL, reqBody)
		if err != nil {
			t.Fatalf("%s: building request: %v", name, err)
		}
		for k, v := range headers {
			req.Header.Set(k, v)
		}
		if s.Request.Body != nil && s.Request.Body.JSON != nil && req.Header.Get("Content-Type") == "" {
			req.Header.Set("Content-Type", "application/json")
			headers["Content-Type"] = "application/json"
		}
		if s.Request.ExpectContinue {
			req.Header.Set("Expect", "100-continue")
		}
		if a.Key != nil {
			key := fixtureKey(t, fx, a.Key.Key)
			sigReq.SignedHeaders = a.Key.SignedHeaders
			ts := time.Now().Unix() + a.Key.TimeOffsetSeconds
			authz := sig.Authorization(sigReq, key, ts)
			if a.Key.CorruptSignature {
				authz = corruptLastHex(authz)
			}
			req.Header.Set("X-Tunna-Date", strconv.FormatInt(ts, 10))
			req.Header.Set("Authorization", authz)
		}

		resp, err := srv.Client().Do(req)
		if err != nil {
			t.Fatalf("%s: %v", name, err)
		}
		respBody, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			t.Fatalf("%s: reading response: %v", name, err)
		}

		// Captures apply inside the expectation too, so a later step can
		// expect the id an earlier one created.
		if s.Expect.JSON != nil {
			s.Expect.JSON = json.RawMessage(subJSON(string(s.Expect.JSON)))
		}
		check(t, name, s.Expect, resp, respBody, statusOf)

		for key, from := range s.Capture {
			val, isNumber, err := capture(from, resp, respBody)
			if err != nil {
				t.Fatalf("%s: capture %s: %v", name, key, err)
			}
			captured[key] = val
			if isNumber {
				numbers[key] = val
			}
		}
	}
}

func check(t *testing.T, name string, e expect, resp *http.Response, body []byte, statusOf map[string]int) {
	wantStatus := e.Status
	if e.Error != nil {
		code := *e.Error
		st, ok := statusOf[code]
		if !ok {
			t.Fatalf("%s: expects error %q which is not in errors.json", name, code)
		}
		if wantStatus == nil {
			wantStatus = &st
		}
		// Every wire error names its code in a header too (wire.md 9), so
		// a HEAD, which has no body, still says what went wrong.
		if got := resp.Header.Get("X-Tunna-Error"); got != code {
			t.Errorf("%s: X-Tunna-Error = %q, want %q", name, got, code)
		}
		if resp.Request.Method != http.MethodHead {
			var eb struct {
				Error struct {
					Code string `json:"code"`
				} `json:"error"`
			}
			if err := json.Unmarshal(body, &eb); err != nil {
				t.Errorf("%s: error body is not JSON: %v\n%s", name, err, body)
			} else if eb.Error.Code != code {
				t.Errorf("%s: error code = %q, want %q", name, eb.Error.Code, code)
			}
		}
	} else if got := resp.Header.Get("X-Tunna-Error"); got != "" && resp.StatusCode < 400 {
		t.Errorf("%s: X-Tunna-Error = %q on a %d response", name, got, resp.StatusCode)
	}
	if wantStatus != nil && resp.StatusCode != *wantStatus {
		t.Errorf("%s: status = %d, want %d\n%s", name, resp.StatusCode, *wantStatus, body)
	}

	for h, raw := range e.Headers {
		var exact string
		if err := json.Unmarshal(raw, &exact); err == nil {
			if got := resp.Header.Get(h); got != exact {
				t.Errorf("%s: header %s = %q, want %q", name, h, got, exact)
			}
			continue
		}
		var p struct {
			Present bool `json:"present"`
		}
		if err := json.Unmarshal(raw, &p); err != nil {
			t.Fatalf("%s: bad header expectation for %s", name, h)
		}
		if _, ok := resp.Header[http.CanonicalHeaderKey(h)]; ok != p.Present {
			t.Errorf("%s: header %s present = %v, want %v", name, h, ok, p.Present)
		}
	}

	if e.JSON != nil || len(e.JSONPresent) > 0 || len(e.JSONAbsent) > 0 {
		var got any
		if err := json.Unmarshal(body, &got); err != nil {
			t.Errorf("%s: body is not JSON: %v\n%s", name, err, body)
		} else {
			if e.JSON != nil {
				var want any
				if err := json.Unmarshal(e.JSON, &want); err != nil {
					t.Fatalf("%s: bad json expectation: %v", name, err)
				}
				if diff := subset(want, got, ""); diff != "" {
					t.Errorf("%s: %s\nbody: %s", name, diff, body)
				}
			}
			for _, ptr := range e.JSONPresent {
				if _, err := pointer(got, ptr); err != nil {
					t.Errorf("%s: %s should be present: %v\nbody: %s", name, ptr, err, body)
				}
			}
			for _, ptr := range e.JSONAbsent {
				if _, err := pointer(got, ptr); err == nil {
					t.Errorf("%s: %s should be absent", name, ptr)
				}
			}
		}
	}

	if e.BodySHA256 != nil {
		sum := sha256.Sum256(body)
		if got := hex.EncodeToString(sum[:]); got != *e.BodySHA256 {
			t.Errorf("%s: body sha256 = %s, want %s", name, got, *e.BodySHA256)
		}
	}
	if e.BodyLength != nil && int64(len(body)) != *e.BodyLength {
		t.Errorf("%s: body length = %d, want %d", name, len(body), *e.BodyLength)
	}
	if e.BodyEmpty != nil && (len(body) == 0) != *e.BodyEmpty {
		t.Errorf("%s: body empty = %v, want %v", name, len(body) == 0, *e.BodyEmpty)
	}
	if e.BodyBytes != nil {
		if !bytes.Equal(body, fixtures.Generate(e.BodyBytes.Seed, e.BodyBytes.Length)) {
			t.Errorf("%s: body does not equal generated bytes (seed %d, length %d)", name, e.BodyBytes.Seed, e.BodyBytes.Length)
		}
	}
}

// --- helpers ---

func mustRead(t *testing.T, path string) string {
	t.Helper()
	b, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("reading %s: %v", path, err)
	}
	return string(b)
}

func loadFixtures(t *testing.T) fixtures.File {
	fx, err := fixtures.Load(filepath.Join(specDir, "conformance", "fixtures.json"))
	if err != nil {
		t.Fatalf("fixtures.json: %v", err)
	}
	return fx
}

func fixtureKey(t *testing.T, fx fixtures.File, name string) sig.Key {
	t.Helper()
	k, ok := fx.SigKey(name)
	if !ok {
		t.Fatalf("no key fixture %q", name)
	}
	return k
}

// corruptLastHex flips the final hex digit of a signature so it is
// well-formed but wrong.
func corruptLastHex(s string) string {
	last := s[len(s)-1]
	if last == '0' {
		return s[:len(s)-1] + "1"
	}
	return s[:len(s)-1] + "0"
}

func loadErrorTable(t *testing.T) map[string]int {
	var tbl errorTable
	if err := json.Unmarshal([]byte(mustRead(t, filepath.Join(specDir, "errors.json"))), &tbl); err != nil {
		t.Fatalf("errors.json: %v", err)
	}
	m := map[string]int{}
	for _, e := range tbl.Errors {
		m[e.Code] = e.Status
	}
	return m
}

// encode is the RFC 3986 unreserved-set percent-encoder from wire.md section 1.
func encode(s string) string {
	const hexDigits = "0123456789ABCDEF"
	var b strings.Builder
	for _, c := range []byte(s) {
		switch {
		case 'A' <= c && c <= 'Z', 'a' <= c && c <= 'z', '0' <= c && c <= '9', c == '-', c == '.', c == '_', c == '~':
			b.WriteByte(c)
		default:
			b.WriteByte('%')
			b.WriteByte(hexDigits[c>>4])
			b.WriteByte(hexDigits[c&15])
		}
	}
	return b.String()
}

func substitute(s string, vars map[string]string) string {
	for k, v := range vars {
		s = strings.ReplaceAll(s, "{{"+k+"}}", v)
	}
	return s
}

// subset reports the first place where want is not a subset of got.
func subset(want, got any, at string) string {
	switch w := want.(type) {
	case map[string]any:
		g, ok := got.(map[string]any)
		if !ok {
			return fmt.Sprintf("%s: want object, got %T", pathOr(at), got)
		}
		for k, wv := range w {
			gv, ok := g[k]
			if !ok {
				return fmt.Sprintf("%s/%s: missing", at, k)
			}
			if d := subset(wv, gv, at+"/"+k); d != "" {
				return d
			}
		}
		return ""
	case []any:
		g, ok := got.([]any)
		if !ok {
			return fmt.Sprintf("%s: want array, got %T", pathOr(at), got)
		}
		if len(g) != len(w) {
			return fmt.Sprintf("%s: array length %d, want %d", pathOr(at), len(g), len(w))
		}
		for i := range w {
			if d := subset(w[i], g[i], fmt.Sprintf("%s/%d", at, i)); d != "" {
				return d
			}
		}
		return ""
	default:
		if !reflect.DeepEqual(want, got) {
			return fmt.Sprintf("%s: got %v, want %v", pathOr(at), got, want)
		}
		return ""
	}
}

func pathOr(at string) string {
	if at == "" {
		return "(root)"
	}
	return at
}

// pointer resolves an RFC 6901 JSON pointer against a decoded value.
func pointer(v any, ptr string) (any, error) {
	if ptr == "" {
		return v, nil
	}
	if !strings.HasPrefix(ptr, "/") {
		return nil, fmt.Errorf("pointer %q must start with /", ptr)
	}
	for _, tok := range strings.Split(ptr[1:], "/") {
		tok = strings.ReplaceAll(strings.ReplaceAll(tok, "~1", "/"), "~0", "~")
		switch cur := v.(type) {
		case map[string]any:
			next, ok := cur[tok]
			if !ok {
				return nil, fmt.Errorf("%q not found", tok)
			}
			v = next
		case []any:
			i, err := strconv.Atoi(tok)
			if err != nil || i < 0 || i >= len(cur) {
				return nil, fmt.Errorf("index %q out of range", tok)
			}
			v = cur[i]
		default:
			return nil, fmt.Errorf("cannot descend into %T at %q", v, tok)
		}
	}
	return v, nil
}

// capture reads one value from the response as text and reports whether it
// was a JSON number.
func capture(from string, resp *http.Response, body []byte) (string, bool, error) {
	if h, ok := strings.CutPrefix(from, "header:"); ok {
		return resp.Header.Get(h), false, nil
	}
	var v any
	if err := json.Unmarshal(body, &v); err != nil {
		return "", false, err
	}
	val, err := pointer(v, from)
	if err != nil {
		return "", false, err
	}
	switch x := val.(type) {
	case string:
		return x, false, nil
	case float64:
		return strconv.FormatFloat(x, 'f', -1, 64), true, nil
	default:
		return fmt.Sprint(x), false, nil
	}
}

File: sdk/typescript/test/conformance.test.ts
// Replays spec/conformance/cases/*.json against cmd/tunna-fixtures with the
// SDK's own encoder and signer (ADR-0010). A port of the Go runner in
// internal/httpapi/conformance_test.go: same file shapes, same checks.
//
// One difference the platform forces: fetch cannot send Expect:
// 100-continue, so expect_continue steps send the body eagerly. The server
// answers the same either way; only the bytes on the wire differ.

import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { createHash } from "node:crypto";
import { readdirSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { encodePath, encodeQuery } from "../src/encode.ts";
import { SPEC_VERSION } from "../src/errors.generated.ts";
import { authorization, HEADER_DATE, presignQuery, type Key, type Request as SigRequest } from "../src/sign.ts";
import { FixtureServer } from "./fixture-server.ts";
import { generate, loadSpec, specVersion } from "./vectors.ts";

// --- spec file shapes (subset of the schemas, enough to run) ---

type Pair = [string, string];
type Gen = { seed: number; length: number };

type Auth =
  | "none"
  | { key: string; signed_headers?: string[]; time_offset_seconds?: number; corrupt_signature?: boolean }
  | { presign: { key: string; expires_in_seconds: number; signed_headers?: string[] } };

type Body = { json?: unknown; text?: string; hex?: string; bytes?: Gen };

type Expect = {
  status?: number;
  error?: string;
  headers?: Record<string, string | { present: boolean }>;
  json?: unknown;
  json_present?: string[];
  json_absent?: string[];
  body_sha256?: string;
  body_length?: number;
  body_empty?: boolean;
  body_bytes?: Gen;
};

type Step = {
  name?: string;
  request: { method: string; path: string[]; query?: Pair[]; headers?: Record<string, string>; auth?: Auth; body?: Body; expect_continue?: boolean };
  expect: Expect;
  capture?: Record<string, string>;
};

type Case = { name: string; description?: string; fixtures?: string[]; steps: Step[] };
type CaseFile = { spec: string; cases: Case[] };

type Fixtures = {
  keys: Record<string, { id: string; secret: string }>;
  buckets: Record<string, unknown>;
};

const fx = loadSpec<Fixtures>("conformance/fixtures.json");
const statusOf = new Map(loadSpec<{ errors: { code: string; status: number }[] }>("errors.json").errors.map((e) => [e.code, e.status]));
const casesDir = new URL("../../../spec/conformance/cases/", import.meta.url);

function fixtureKey(name: string): Key {
  const k = fx.keys[name];
  if (!k) throw new Error(`no key fixture ${name}`);
  return { id: k.id, secret: k.secret };
}

function unsupported(c: Case): string | undefined {
  for (const name of c.fixtures ?? []) {
    if (!(name in fx.keys) && !(name in fx.buckets)) return `unknown fixture ${name}`;
  }
  return undefined;
}

// --- the runner ---

let server: FixtureServer;

beforeAll(async () => {
  server = await FixtureServer.start();
});

afterAll(() => {
  server?.stop();
});

test("the fixture server implements this package's spec version", async () => {
  const res = await fetch(`${server.url}/-/version`);
  const body = (await res.json()) as { spec: string };
  expect(body.spec).toBe(SPEC_VERSION);
});

for (const file of readdirSync(casesDir).filter((f) => f.endsWith(".json")).sort()) {
  const cf = loadSpec<CaseFile>(`conformance/cases/${file}`);

  describe(file, () => {
    test("declares the current spec version", () => {
      expect(cf.spec).toBe(specVersion());
    });

    for (const c of cf.cases) {
      const reason = unsupported(c);
      const run = reason ? test.skip : test;
      run(c.name, async () => {
        // A fresh server per case: no case can poison another, and a case
        // that mutates a fixture is still a valid case.
        await server.reset();
        await runCase(c);
      });
    }
  });
}

async function runCase(c: Case): Promise<void> {
  const captured = new Map<string, string>();
  const sub = (s: string) => substitute(s, captured);

  for (const [i, s] of c.steps.entries()) {
    const name = s.name ?? `step-${i + 1}`;
    const at = (msg: string) => `${c.name} / ${name}: ${msg}`;

    // Decoded inputs, after substitution.
    const segments = s.request.path.map(sub);
    const query: Pair[] = (s.request.query ?? []).map(([k, v]) => [sub(k), sub(v)]);
    const headers: Record<string, string> = {};
    for (const [k, v] of Object.entries(s.request.headers ?? {})) headers[k] = sub(v);

    // Body
    let body: Uint8Array<ArrayBuffer> | string | undefined;
    if (s.request.body) {
      const b = s.request.body;
      if (b.json !== undefined) {
        body = sub(JSON.stringify(b.json));
        if (!Object.keys(headers).some((h) => h.toLowerCase() === "content-type")) headers["Content-Type"] = "application/json";
      } else if (b.text !== undefined) {
        body = sub(b.text);
      } else if (b.hex !== undefined) {
        body = Uint8Array.from(Buffer.from(b.hex, "hex"));
      } else if (b.bytes !== undefined) {
        body = generate(b.bytes.seed, b.bytes.length);
      }
    }

    const now = Math.floor(Date.now() / 1000);
    const auth = s.request.auth ?? "none";
    const sigReq: SigRequest = { method: s.request.method, path: segments, query, headers };
    let url = server.url + encodePath(segments);

    if (auth !== "none" && "presign" in auth) {
      const p = auth.presign;
      if (p.signed_headers) sigReq.signedHeaders = p.signed_headers;
      url += "?" + (await presignQuery(sigReq, fixtureKey(p.key), now + p.expires_in_seconds));
    } else if (query.length > 0) {
      url += "?" + encodeQuery(query);
    }

    if (auth !== "none" && "key" in auth) {
      if (auth.signed_headers) sigReq.signedHeaders = auth.signed_headers;
      const ts = now + (auth.time_offset_seconds ?? 0);
      let authz = await authorization(sigReq, fixtureKey(auth.key), ts);
      if (auth.corrupt_signature) authz = corruptLastHex(authz);
      headers[HEADER_DATE] = String(ts);
      headers["Authorization"] = authz;
    }

    const res = await fetch(url, { method: s.request.method, headers, body: body ?? null });
    const respBody = new Uint8Array(await res.arrayBuffer());

    check(at, s.expect, res, respBody, sub);

    for (const [key, from] of Object.entries(s.capture ?? {})) {
      captured.set(key, capture(from, res, respBody));
    }
  }
}

function check(at: (m: string) => string, e: Expect, res: Response, body: Uint8Array, sub: (s: string) => string): void {
  const text = () => new TextDecoder().decode(body);
  let wantStatus = e.status;

  if (e.error !== undefined) {
    const st = statusOf.get(e.error);
    if (st === undefined) throw new Error(at(`expects error ${e.error} which is not in errors.json`));
    wantStatus ??= st;
    const eb = parseJSON(body);
    expect(eb, at(`error body is not JSON: ${text()}`)).not.toBeUndefined();
    expect((eb as { error?: { code?: string } })?.error?.code, at("error code")).toBe(e.error);
  }
  if (wantStatus !== undefined) {
    expect(res.status, at(`status; body: ${text()}`)).toBe(wantStatus);
  }

  for (const [h, want] of Object.entries(e.headers ?? {})) {
    if (typeof want === "string") {
      expect(res.headers.get(h) ?? "", at(`header ${h}`)).toBe(want);
    } else {
      expect(res.headers.has(h), at(`header ${h} present`)).toBe(want.present);
    }
  }

  if (e.json !== undefined || e.json_present?.length || e.json_absent?.length) {
    const got = parseJSON(body);
    expect(got, at(`body is not JSON: ${text()}`)).not.toBeUndefined();
    if (e.json !== undefined) {
      // Captures apply inside the expectation too.
      const want = JSON.parse(sub(JSON.stringify(e.json)));
      const diff = subset(want, got, "");
      if (diff) throw new Error(at(`${diff}\nbody: ${text()}`));
    }
    for (const ptr of e.json_present ?? []) {
      if (pointer(got, ptr) === undefined) throw new Error(at(`${ptr} should be present\nbody: ${text()}`));
    }
    for (const ptr of e.json_absent ?? []) {
      if (pointer(got, ptr) !== undefined) throw new Error(at(`${ptr} should be absent`));
    }
  }

  if (e.body_sha256 !== undefined) {
    expect(createHash("sha256").update(body).digest("hex"), at("body sha256")).toBe(e.body_sha256);
  }
  if (e.body_length !== undefined) {
    expect(body.length, at("body length")).toBe(e.body_length);
  }
  if (e.body_empty !== undefined) {
    expect(body.length === 0, at("body empty")).toBe(e.body_empty);
  }
  if (e.body_bytes !== undefined) {
    expect(Buffer.from(body).equals(generate(e.body_bytes.seed, e.body_bytes.length)), at("body equals generated bytes")).toBe(true);
  }
}

// --- helpers ---

function parseJSON(body: Uint8Array): unknown {
  try {
    return JSON.parse(new TextDecoder().decode(body));
  } catch {
    return undefined;
  }
}

function substitute(s: string, vars: Map<string, string>): string {
  for (const [k, v] of vars) s = s.replaceAll(`{{${k}}}`, v);
  return s;
}

function corruptLastHex(s: string): string {
  return s.slice(0, -1) + (s.endsWith("0") ? "1" : "0");
}

/** The first place where want is not a subset of got, or "" when it is. */
function subset(want: unknown, got: unknown, at: string): string {
  const here = at || "/";
  if (Array.isArray(want)) {
    if (!Array.isArray(got)) return `${here}: want array, got ${typeof got}`;
    if (got.length !== want.length) return `${here}: array length ${got.length}, want ${want.length}`;
    for (let i = 0; i < want.length; i++) {
      const d = subset(want[i], got[i], `${at}/${i}`);
      if (d) return d;
    }
    return "";
  }
  if (want !== null && typeof want === "object") {
    if (got === null || typeof got !== "object" || Array.isArray(got)) return `${here}: want object, got ${typeof got}`;
    for (const [k, wv] of Object.entries(want)) {
      if (!(k in got)) return `${at}/${k}: missing`;
      const d = subset(wv, (got as Record<string, unknown>)[k], `${at}/${k}`);
      if (d) return d;
    }
    return "";
  }
  return Object.is(want, got) ? "" : `${here}: got ${JSON.stringify(got)}, want ${JSON.stringify(want)}`;
}

/** RFC 6901 JSON pointer; undefined when it does not resolve. */
function pointer(v: unknown, ptr: string): unknown {
  if (ptr === "") return v;
  if (!ptr.startsWith("/")) return undefined;
  let cur = v;
  for (const raw of ptr.slice(1).split("/")) {
    const tok = raw.replaceAll("~1", "/").replaceAll("~0", "~");
    if (Array.isArray(cur)) {
      const i = Number(tok);
      if (!Number.isInteger(i) || i < 0 || i >= cur.length) return undefined;
      cur = cur[i];
    } else if (cur !== null && typeof cur === "object") {
      if (!(tok in cur)) return undefined;
      cur = (cur as Record<string, unknown>)[tok];
    } else {
      return undefined;
    }
  }
  return cur;
}

function capture(from: string, res: Response, body: Uint8Array): string {
  if (from.startsWith("header:")) return res.headers.get(from.slice("header:".length)) ?? "";
  const v = parseJSON(body);
  if (v === undefined) throw new Error(`capture ${from}: body is not JSON`);
  const val = pointer(v, from);
  if (val === undefined) throw new Error(`capture ${from}: not found`);
  return typeof val === "string" ? val : JSON.stringify(val);
}
